File: examples/CBuilder/Simple/GUIControls/TreeView/fTreeViewMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "fTreeViewMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BoldAbstractListHandle"
#pragma link "BoldAbstractModel"
#pragma link "BoldActions"
#pragma link "BoldCursorHandle"
#pragma link "BoldDBActions"
#pragma link "BoldExpressionHandle"
#pragma link "BoldGrid"
#pragma link "BoldHandle"
#pragma link "BoldHandleAction"
#pragma link "BoldHandles"
#pragma link "BoldListBox"
#pragma link "BoldListHandle"
#pragma link "BoldModel"
#pragma link "BoldNavigator"
#pragma link "BoldNavigatorDefs"
#pragma link "BoldPersistenceHandle"
#pragma link "BoldPersistenceHandleDB"
#pragma link "BoldReferenceHandle"
#pragma link "BoldRootedHandles"
#pragma link "BoldSubscription"
#pragma link "BoldSystemHandle"
#pragma link "BoldTreeView"
#pragma link "BoldEdit"
#pragma link "BoldGUI"
#pragma link "BoldAbstractListHandle"
#pragma link "BoldAbstractModel"
#pragma link "BoldActions"
#pragma link "BoldCursorHandle"
#pragma link "BoldDBActions"
#pragma link "BoldEdit"
#pragma link "BoldExpressionHandle"
#pragma link "BoldGrid"
#pragma link "BoldHandle"
#pragma link "BoldHandleAction"
#pragma link "BoldHandles"
#pragma link "BoldListBox"
#pragma link "BoldListHandle"
#pragma link "BoldModel"
#pragma link "BoldNavigator"
#pragma link "BoldNavigatorDefs"
#pragma link "BoldPersistenceHandle"
#pragma link "BoldPersistenceHandleBDE"
#pragma link "BoldPersistenceHandleDB"
#pragma link "BoldReferenceHandle"
#pragma link "BoldRootedHandles"
#pragma link "BoldSubscription"
#pragma link "BoldSystem"
#pragma link "BoldTreeView"
#pragma link "BoldSystemHandle"
#pragma link "BoldAbstractDatabaseAdapter"
#pragma link "BoldAbstractPersistenceHandleDB"
#pragma link "BoldDatabaseAdapterIB"
#pragma link "BoldIBDatabaseAction"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCloseQuery(TObject *Sender, bool &CanClose)
{
  CanClose = true;
  if (BoldSystemHandle1->Active)
    if (BoldSystemHandle1->System->DirtyObjects->Count > 0)
      switch (MessageDlg("There are dirty objects. save them before exit?",
      mtConfirmation, TMsgDlgButtons() << mbYes << mbNo << mbCancel, 0))
      {
        case mrYes: BoldSystemHandle1->System->UpdateDatabase(); break;
        case mrNo: BoldSystemHandle1->System->Discard(); break;
        case mrCancel: CanClose = false;
      }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCreate(TObject *Sender)
{
  Height = 508;
  Width = 641;
  pgcTreeViews->ActivePage = tsBasicTree;
  CurrentTreeView = btrvBasic;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnCurrentAsRootClick(TObject *Sender)
{
  brhTreeRoot->Value = CurrentTreeView->CurrentElement;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnShowAllClick(TObject *Sender)
{
  brhTreeRoot->Value = BoldSystemHandle1->System->ClassByExpressionName["MTB"];
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OpenSystemClicked(TObject *Sender)
{
  if (BoldSystemHandle1->Active == true)
    BoldActivateSystemAction1SystemClosed(Sender);
  else
    BoldActivateSystemAction1SystemOpened(Sender);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::BoldActivateSystemAction1SystemClosed(TObject *Sender)
{
    if (btnShowAll)
      btnShowAll->Enabled = BoldSystemHandle1->Active;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::BoldActivateSystemAction1SystemOpened(TObject *Sender)
{
  if (BoldSystemHandle1->System->ClassByExpressionName["BikeFrame"]->Count < 1)
    PopulateClass("BikeFrame", "Name", BoldSystemHandle1->System->ClassByExpressionName["BikeFrame"]);
  if (BoldSystemHandle1->System->ClassByExpressionName["Gear"]->Count < 1)
    PopulateClass("Gear", "Model", BoldSystemHandle1->System->ClassByExpressionName["Gear"]);
  if (BoldSystemHandle1->System->ClassByExpressionName["Brake"]->Count < 1)
    PopulateClass("Brake", "Model", BoldSystemHandle1->System->ClassByExpressionName["Brake"]);
  if (BoldSystemHandle1->System->ClassByExpressionName["Wheel"]->Count < 1)
    PopulateClass("Wheel", "Model", BoldSystemHandle1->System->ClassByExpressionName["Wheel"]);
  if (BoldSystemHandle1->System->ClassByExpressionName["MTB"]->Count < 1)
    PopulateClass("MTB", "Name", BoldSystemHandle1->System->ClassByExpressionName["MTB"]);
  if (BoldSystemHandle1->System->DirtyObjects->Count > 0)
    BoldSystemHandle1->System->UpdateDatabase();

  brhTreeRoot->Value = BoldSystemHandle1->System->ClassByExpressionName["MTB"];
  btnShowAll->Enabled = BoldSystemHandle1->Active;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::imgUnlinkDragDrop(TObject *Sender, TObject *Source, int X, int Y)
{
  TBoldObject *ObjectToUnlink;
  ObjectToUnlink = BoldGUIHandler()->DraggedObjects->BoldObjects[0];
  if ((dynamic_cast<TBikeFrame*>(ObjectToUnlink)) != 0)
    ((TMTB*)DragContext)->BuiltAround = NULL;
  else
    ((TMTB*)DragContext)->ConsistsOf->Remove(ObjectToUnlink);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::imgUnlinkDragOver(TObject *Sender, TObject *Source, int X, int Y, TDragState State, bool &Accept)
{
  Accept = false;
  if ((BoldGUIHandler()->DraggedObjects->Count == 1) && (DragContext))
    Accept = true;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btrvEnhancedDragDrop(TObject *Sender, TObject *Source, int X, int Y)
{
  TBoldObject *ObjectToLink;
  ObjectToLink = BoldGUIHandler()->DraggedObjects->BoldObjects[0];
  if ((dynamic_cast<TParts*>(ObjectToLink)) != 0)
    TargetObject->ConsistsOf->Add(dynamic_cast<TParts*>(ObjectToLink));
  else
    TargetObject->BuiltAround = dynamic_cast<TBikeFrame*>(ObjectToLink);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btrvEnhancedDragOver(TObject *Sender, TObject *Source, int X, int Y, TDragState State, bool &Accept)
{
  TBoldTreeView *TreeView;
  TBoldElement *TargetElement;
  Accept = false;

  TreeView = (TBoldTreeView*)Sender;
  TargetElement = TreeView->GetElementAt(X, Y);

  TBoldObjectList* objectlist;
  objectlist = BoldGUIHandler()->DraggedObjects;

  if (TargetElement)
  {
    if ((objectlist->Count == 1) && ((dynamic_cast<TMTB*>(TargetElement)) != 0)
      && ((dynamic_cast<TMTB*>(objectlist->BoldObjects[0])) == 0 ))
    {
      TargetObject = dynamic_cast<TMTB*>(TargetElement);
      Accept = ( ((dynamic_cast<TParts*>(objectlist->BoldObjects[0])) != 0)
                || ((dynamic_cast<TBikeFrame*>(objectlist->BoldObjects[0])) != 0) );
    }
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::pgcTreeViewsChange(TObject *Sender)
{
  if (pgcTreeViews->ActivePage == tsBasicTree)
    CurrentTreeView = btrvBasic;
  else
    CurrentTreeView = btrvEnhanced;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::PopulateClass(AnsiString AClassName, AnsiString APropName, TBoldObjectList *AList)
{
  TStrings *NameStrings = new TStringList();
  int i;
  TBoldElement *AElement;

  NameStrings->LoadFromFile(AClassName + ".txt");
  for (i = 0; i < NameStrings->Count; i++)
  {
    AElement = AList->AddNew();
    ((TBoldObject*)AElement)->BoldMemberByExpressionName[APropName]->AsString = NameStrings->Strings[i];
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btrvEnhancedStartDrag(TObject *Sender, TDragObject *&DragObject)
{
  TBoldElement *ABoldElement;
  DragContext = NULL;

  if ( (CurrentTreeView->Selected) && (CurrentTreeView->Selected->Parent) )
  {
    ABoldElement = ((TBoldTreeNode*)(CurrentTreeView->Selected->Parent))->Follower->Element;
    if ( (dynamic_cast<TMTB*>(ABoldElement)) != 0)
      DragContext = (TBoldObject*)ABoldElement;
  }
}
//---------------------------------------------------------------------------

File: examples/Delphi/Simple/GUIControls/TreeView/fTreeViewMain.pas
unit fTreeViewMain;

interface

uses
  SysUtils,
  Classes,
  Graphics,
  Controls,
  Forms,
  Dialogs,
  ComCtrls,
  ImgList,
  ActnList,
  StdCtrls,
  ExtCtrls,
  Grids,
  BoldHandle,
  BoldHandles,
  BoldSystemHandle,
  BoldExpressionHandle,
  BoldRootedHandles,
  BoldReferenceHandle,
  BoldAbstractListHandle,
  BoldCursorHandle,
  BoldListHandle,
  BoldPersistenceHandle,
  BoldPersistenceHandleDB,
  BoldSubscription,
  BoldModel,
  BoldNavigator,
  BoldTreeView,
  BoldEdit,
  BoldListBox,
  BoldGrid,
  BoldHandleAction,
  BoldActions,
  BoldDBActions, BoldAbstractModel, BoldNavigatorDefs,
  BoldIBDatabaseAction, DB, IBDatabase, BoldAbstractDatabaseAdapter,
  BoldDatabaseAdapterIB, BoldAbstractPersistenceHandleDB;

type
  TForm1 = class(TForm)
    BoldModel1: TBoldModel;
    BoldSystemHandle1: TBoldSystemHandle;
    BoldSystemTypeInfoHandle1: TBoldSystemTypeInfoHandle;
    blhMTB: TBoldListHandle;
    blhAllFrames: TBoldListHandle;
    blhAllBrakes: TBoldListHandle;
    blhAllGears: TBoldListHandle;
    blhAllWheels: TBoldListHandle;
    blhComponents: TBoldListHandle;
    behFrame: TBoldExpressionHandle;
    brhTreeRoot: TBoldReferenceHandle;
    btnShowAll: TButton;
    pgcTreeViews: TPageControl;
    tsBasicTree: TTabSheet;
    btrvBasic: TBoldTreeView;
    tsEnhancedTree: TTabSheet;
    btrvEnhanced: TBoldTreeView;
    imlTreeNodeIcons: TImageList;
    grpFrames: TGroupBox;
    BoldGrid2: TBoldGrid;
    BoldNavigator2: TBoldNavigator;
    grpBrakes: TGroupBox;
    grpGears: TGroupBox;
    GroupBox3: TGroupBox;
    BoldGrid3: TBoldGrid;
    BoldNavigator3: TBoldNavigator;
    BoldGrid4: TBoldGrid;
    BoldNavigator4: TBoldNavigator;
    BoldGrid5: TBoldGrid;
    BoldNavigator5: TBoldNavigator;
    BoldGrid1: TBoldGrid;
    lblFrame: TLabel;
    BoldEdit1: TBoldEdit;
    lblComponents: TLabel;
    BoldListBox1: TBoldListBox;
    BoldNavigator1: TBoldNavigator;
    lblMountainbike: TLabel;
    btnCurrentAsRoot: TButton;
    Panel1: TPanel;
    imgUnlink: TImage;
    ActionList1: TActionList;
    BoldActivateSystemAction1: TBoldActivateSystemAction;
    Button1: TButton;
    Button2: TButton;
    BoldPersistenceHandleDB1: TBoldPersistenceHandleDB;
    BoldDatabaseAdapterIB1: TBoldDatabaseAdapterIB;
    IBDatabase1: TIBDatabase;
    BoldIBDatabaseAction1: TBoldIBDatabaseAction;
    procedure btnShowAllClick(Sender: TObject);
    procedure btrvEnhancedDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
    procedure btrvEnhancedDragDrop(Sender, Source: TObject; X,
      Y: Integer);
    procedure FormCreate(Sender: TObject);
    procedure btnCurrentAsRootClick(Sender: TObject);
    procedure btrvEnhancedStartDrag(Sender: TObject;
      var DragObject: TDragObject);
    procedure pgcTreeViewsChange(Sender: TObject);
    procedure imgUnlinkDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
    procedure imgUnlinkDragDrop(Sender, Source: TObject; X, Y: Integer);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure BoldActivateSystemAction1SystemOpened(Sender: TObject);
    procedure BoldActivateSystemAction1SystemClosed(Sender: TObject);
  private
    { Private declarations }
    procedure PopulateClass(AClassname, APropName: string; AListHandle: TBoldListHandle);
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

uses
  BoldElements,
  BoldGUI,
  BoldSystem,
  TreeViewExampleClasses;

var
  DragContext: TBoldObject;
  TargetObject: TMTB;
  CurrentTreeView: TBoldTreeView;

{$R *.DFM}

// Show all MTB in treeview.
procedure TForm1.btnShowAllClick(Sender: TObject);
begin
    brhTreeRoot.value := BoldSystemHandle1.System.ClassByExpressionName['MTB'];
end;

// Accept drop if we recieve only one object that is not a MTB.
procedure TForm1.btrvEnhancedDragOver(Sender, Source: TObject; X,
  Y: Integer; State: TDragState; var Accept: Boolean);
var
  TargetElement: TBoldElement;
begin
  Accept := False;
  TargetElement := (sender as TBoldTreeView).GetElementAt(X, Y);
  if (BoldGuiHandler.DraggedObjects.Count = 1) and
     (TargetElement is TMTB) and 
     not (BoldGuiHandler.DraggedObjects[0] is TMTB) then
  begin
    TargetObject := TargetElement as TMTB;
    Accept := (BoldGuiHandler.DraggedObjects[0] is TParts) or
              (BoldGuiHandler.DraggedObjects[0] is TFrame);
  end;
end;

// Replace frame or add new components.
procedure TForm1.btrvEnhancedDragDrop(Sender, Source: TObject; X,
  Y: Integer);
var
  ObjectToLink: TBoldObject;
begin
  ObjectToLink := BoldGuiHandler.DraggedObjects[0];
  if ObjectToLink is TParts then
    TargetObject.ConsistsOf.Add(ObjectToLink as TParts)
  else
    TargetObject.BuiltAround := ObjectToLink as TFrame;
end;

// Set size of form and set current treeview.
procedure TForm1.FormCreate(Sender: TObject);
begin
  Height := 508;
  Width  := 641;
  pgcTreeViews.ActivePage := tsBasicTree;
  CurrentTreeView := btrvBasic;
end;

// Set current object to root for treviews.
procedure TForm1.btnCurrentAsRootClick(Sender: TObject);
begin
  brhTreeRoot.Value := CurrentTreeView.CurrentElement;
end;

// Set DragContext if the current object has a MTB parent.
procedure TForm1.btrvEnhancedStartDrag(Sender: TObject;
  var DragObject: TDragObject);
var
  ABoldElement: TBoldElement;
begin
  DragContext := nil;
  if Assigned(CurrentTreeView.Selected) and assigned(CurrentTreeView.Selected.Parent) then
  begin
    ABoldElement := (CurrentTreeView.Selected.Parent as TBoldTreeNode).Follower.Element;
    if ABoldElement is TMTB then
      DragContext := ABoldElement as TBoldObject;
  end;
end;

// Set CurrentTreeView.
procedure TForm1.pgcTreeViewsChange(Sender: TObject);
begin
  if pgcTreeViews.ActivePage = tsBasicTree then
    CurrentTreeView := btrvBasic
  else
    CurrentTreeView := btrvEnhanced;
end;

// Acccept drop if there is only one dragged object and if DragContext is set.
procedure TForm1.imgUnlinkDragOver(Sender, Source: TObject; X, Y: Integer;
  State: TDragState; var Accept: Boolean);
begin
  Accept := False;
  Accept := (BoldGuiHandler.DraggedObjects.Count = 1) and
            (Assigned(DragContext));
end;

// Unlink dragged object from its parent.
procedure TForm1.imgUnlinkDragDrop(Sender, Source: TObject; X, Y: Integer);
var
  ObjectToUnlink: TBoldObject;
begin
  ObjectToUnlink := BoldGuiHandler.DraggedObjects[0];
  if ObjectToUnlink is TFrame then
    (DragContext as TMTB).BuiltAround := nil
  else
    (DragContext as TMTB).ConsistsOf.Remove(ObjectToUnlink);
end;

// Populate a class with a few objects.
procedure TForm1.PopulateClass(AClassName, APropName: string; AListHandle: TBoldListHandle);
var
  NameStrings:  TStrings;
  i: integer;
  AElement: TBoldElement;
begin
  NameStrings := TStringList.Create;
  NameStrings.LoadFromFile(AClassName+ '.txt');
  for i:=0 to NameStrings.Count -1 do
  begin
    AElement := AListHandle.List.AddNew;
    TBoldObject(AElement).BoldMemberByExpressionName[APropName].AsString := NameStrings[i];
  end;
end;

// Check if its OK to close the form.
procedure TForm1.FormCloseQuery(Sender: TObject; var CanClose: Boolean);
begin
  CanClose := True;
  if BoldSystemHandle1.Active then
  if BoldSystemHandle1.System.DirtyObjects.Count > 0 then
    case MessageDlg( 'There are dirty objects. save them before exit?', mtConfirmation, [mbYes, mbNo, mbCancel], 0) of
      mrYes: BoldSystemHandle1.System.UpdateDatabase;
      mrNo: BoldSystemHandle1.System.Discard;
      mrCancel: CanClose := False;
    end;
end;

// Set root expression for treeviews when system opens.
procedure TForm1.BoldActivateSystemAction1SystemOpened(Sender: TObject);
begin
  if blhAllFrames.List.Count < 1 then
    PopulateClass('Frame', 'Name', blhAllFrames);
  if blhAllGears.List.Count < 1 then
    PopulateClass('Gear', 'Model', blhAllGears);
  if blhAllBrakes.List.Count < 1 then
    PopulateClass('Brake', 'Model', blhAllBrakes);
  if blhAllWheels.List.Count < 1 then
    PopulateClass('Wheel', 'Model', blhAllWheels);
  if blhMTB.List.Count < 1 then
    PopulateClass('MTB', 'Name', blhMTB);
  if BoldSystemHandle1.System.DirtyObjects.Count > 0 then
    BoldSystemHandle1.System.UpdateDatabase;
  brhTreeRoot.Value := BoldSystemHandle1.System.ClassByExpressionName['MTB'];
  btnShowAll.Enabled := BoldSystemHandle1.Active;  
end;

procedure TForm1.BoldActivateSystemAction1SystemClosed(Sender: TObject);
begin
  btnShowAll.Enabled := BoldSystemHandle1.Active;
end;

end.
